tailfs/compositefs/compositefs_test.go

```diff
@@ -0,0 +1,507 @@
+			oldName: "/remote2/file2.txt",
+			newName: "/remote2/file3.txt",
+			expectedNewInfo: &shared.StaticFileInfo{
+				Named: "/remote2/file3.txt",
+				Sized: 5,
+				Dir:   false,
+			},
+		},
+	}
+
+	ctx := context.Background()
+	for _, test := range tests {
+		t.Run(test.label, func(t *testing.T) {
+			err := fs.Rename(ctx, test.oldName, test.newName)
+			if test.err != nil {
+				if err == nil || test.err.Error() != err.Error() {
+					t.Errorf("expected error: %v   got: %v", test.err, err)
+				}
+			} else {
+				if err != nil {
+					t.Errorf("unexpected error: %v", err)
+				} else {
+					fi, err := fs.Stat(ctx, test.newName)
+					if err != nil {
+						t.Errorf("unexpected error: %v", err)
+					} else {
+						// Override modTime to avoid having to compare it
+						test.expectedNewInfo.ModdedTime = fi.ModTime()
+						infosEqual(t, test.expectedNewInfo, fi)
+					}
+				}
+			}
+		})
+	}
+}
+
+func createFileSystem(t *testing.T, opts *Opts) (webdav.FileSystem, string, string, *tstest.Clock, func()) {
+	l1, dir1 := startRemote(t)
+	l2, dir2 := startRemote(t)
+
+	// Make some files, use perms 0666 as lowest common denominator that works
+	// on both UNIX and Windows.
+	err := os.WriteFile(filepath.Join(dir1, "file1.txt"), []byte("12345"), 0666)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = os.WriteFile(filepath.Join(dir2, "file2.txt"), []byte("54321"), 0666)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// make some directories
+	err = os.Mkdir(filepath.Join(dir1, "dir1"), 0666)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = os.Mkdir(filepath.Join(dir2, "dir2"), 0666)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if opts == nil {
+		opts = &Opts{}
+	}
+	if opts.Logf == nil {
+		opts.Logf = t.Logf
+	}
+	clock := tstest.NewClock(tstest.ClockOpts{Start: time.Now()})
+	opts.Clock = clock
+
+	fs := New(opts)
+	fs.AddChild(&Child{Name: "remote4", FS: &closeableFS{webdav.Dir(dir2)}})
+	fs.SetChildren(&Child{Name: "remote2", FS: webdav.Dir(dir2)},
+		&Child{Name: "remote3", FS: &closeableFS{webdav.Dir(dir2)}},
+	)
+	fs.AddChild(&Child{Name: "remote1", FS: webdav.Dir(dir1)})
+	fs.RemoveChild("remote3")
+
+	child, ok := fs.GetChild("remote1")
+	if !ok || child == nil {
+		t.Fatal("unable to GetChild(remote1)")
+	}
+	child, ok = fs.GetChild("remote2")
+	if !ok || child == nil {
+		t.Fatal("unable to GetChild(remote2)")
+	}
+	child, ok = fs.GetChild("remote3")
+	if ok || child != nil {
+		t.Fatal("should have been able to GetChild(remote3)")
+	}
+	child, ok = fs.GetChild("remote4")
+	if ok || child != nil {
+		t.Fatal("should have been able to GetChild(remote4)")
+	}
+
+	return fs, dir1, dir2, clock, func() {
+		defer l1.Close()
+		defer os.RemoveAll(dir1)
+		defer l2.Close()
+		defer os.RemoveAll(dir2)
+	}
+}
+
+func stat(t *testing.T, path string) fs.FileInfo {
+	fi, err := os.Stat(path)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return fi
+}
+
+func startRemote(t *testing.T) (net.Listener, string) {
+	dir, err := os.MkdirTemp("", "")
```

Comment:
Consider using `t.TempDir` here, which will clean up the directory when the governing test unwinds.

Also: Probably we should use `t.Cleanup` to wrap up the connections and other resources the webdav package may be using, e.g.,

```go
t.Cleanup(func() {
  lst.Close()
  h.Close()
  s.Shutdown(context.Background())
})
```

or whatever else may be relevant (the above is an example, I didn't check all the corners)

Reply:
> Consider using t.TempDir here, which will clean up the directory when the governing test unwinds.

Ah, that was a leftover from when I wanted to actually look at what was in the directory post-test. Changed.

Reply:
Fixed